appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,691 @@
+ * @param {Genetics.Mouse} mouse The mouse to kill.
+ */
+Genetics.Cage.death = function(mouse) {
+  var index = Genetics.Cage.aliveMice_.indexOf(mouse);
+  if (index > -1) {
+    Genetics.Cage.aliveMice_.splice(index, 1);
+  }
+  delete Genetics.Cage.miceMap_[mouse.id];
+  clearTimeout(Genetics.Cage.lifePidsMap_[mouse.id]);
+  delete Genetics.Cage.lifePidsMap_[mouse.id];
+};
+
+/**
+ * Checks if mouse is alive.
+ * @param {Genetics.Mouse} mouse The mouse to check whether it is alive.
+ * @return {boolean} True if mouse is alive, false otherwise.
+ */
+Genetics.Cage.isAlive = function(mouse) {
+  return (Genetics.Cage.aliveMice_.indexOf(mouse) > -1);
+};
+
+/**
+ * Stops the game if end condition is met.
+ */
+Genetics.Cage.checkForEnd = function() {
+  // Check if there are no mice left.
+  if (Genetics.Cage.aliveMice_.length == 0) {
+    Genetics.Cage.end('NONE_LEFT');
+    return;
+  }
+  // Check which genes have won.
+  var pickFightWinner = Genetics.Cage.aliveMice_[0].pickFightOwner;
+  var mateQuestionWinner = Genetics.Cage.aliveMice_[0].proposeMateOwner;
+  var acceptMateWinner = Genetics.Cage.aliveMice_[0].acceptMateOwner;
+  for (var i = 1; i < Genetics.Cage.aliveMice_.length; i++) {
+    var mouse = Genetics.Cage.aliveMice_[i];
+    if (pickFightWinner && pickFightWinner != mouse.pickFightOwner) {
+      pickFightWinner = null;
+    }
+    if (mateQuestionWinner && mateQuestionWinner != mouse.proposeMateOwner) {
+      mateQuestionWinner = null;
+    }
+    if (acceptMateWinner && acceptMateWinner != mouse.acceptMateOwner) {
+      acceptMateWinner = null;
+    }
+    if (!pickFightWinner && !mateQuestionWinner && !acceptMateWinner) {
+      break;
+    }
+  }
+  // End the game if there is a winner for all functions.
+  if (pickFightWinner && mateQuestionWinner && acceptMateWinner) {
+    Genetics.Cage.end('DOMINATION', pickFightWinner, mateQuestionWinner,
+        acceptMateWinner);
+  }
+  // Check if game has gone on too long.
+  if (Date.now() > Genetics.Cage.endTime_) {
+    Genetics.Cage.end('TIMEOUT', pickFightWinner, mateQuestionWinner,
+        acceptMateWinner);
+  }
+};
+
+/**
+ * Clears queued events for end of simulation.
+ * @param {string} cause The cause of game end.
+ * @param {string=} opt_pickFightWinner The owner of all the pickFight functions
+ * in alive mice.
+ * @param {string=} opt_proposeMateWinner The owner of all the proposeMate
+ * functions in alive mice.
+ * @param {string=} opt_acceptMateWinner The owner of all the acceptMate
+ * functions in alive mice.
+ */
+Genetics.Cage.end = function(cause, opt_pickFightWinner, opt_proposeMateWinner,
+    opt_acceptMateWinner) {
+  // Cancel all queued life simulations.
+  for (var mouseId in Genetics.Cage.lifePidsMap_) {
+    clearTimeout(Genetics.Cage.lifePidsMap_[mouseId]);
+  }
+  var pickFightWinner = opt_pickFightWinner || 'NONE';
+  var proposeMateWinner = opt_proposeMateWinner || 'NONE';
+  var acceptMateWinner = opt_acceptMateWinner || 'NONE';
```

Comment:
Why not delete these three lines and just use null or undefined for indicating there's no winner?  That way there's no risk in the event that an arbitrary ID might happen to be 'NONE'.